play-json-circe/src/main/scala/com/evolutiongaming/util/PlayCirceConversions.scala

```diff
@@ -0,0 +1,50 @@
+package com.evolutiongaming.util
+
+import io.circe._
+import io.circe.syntax._
+import play.api.libs.json.{Json => PlayJson, _}
+
+import scala.util.Try
+
+object PlayCirceConversions {
+
+  private def convert(value: JsValue): Json = {
+    import io.circe.Encoder._
+
+    value match {
+      case JsNull          =>
+        None.asJson
+      case JsTrue          =>
+        true.asJson
+      case JsFalse         =>
+        false.asJson
+      case JsNumber(value) =>
+        value.asJson
+      case JsString(value) =>
+        value.asJson
+      case JsArray(value)  =>
+        value.map(convert).asJson
+      case JsObject(value) =>
+        JsonObject.fromIterable(value.iterator.map { case (k, v) => (k, convert(v)) }.to(Iterable)).asJson
+    }
+  }
+
+  implicit def decoderFromReads[T: Reads]: Decoder[T] = Decoder.decodeJson.emap { json =>
+    Try(PlayJson.parse(json.noSpaces).validate[T]).toEither.left.map(_.getMessage).flatMap {
```

Comment:
imho to much code in single line, could we split or chain lines below?

Reply:
Thanks for pointing this out. It was copy of internal code which I didn't look into.

While re-aligning code I noticed that there is something really weird going on here.
We incur cost of parsing input into AST twice, once by circe, then immediately serialise to string and parse again with play-json.
I'd like to see how can we do it better.

Reply:
correct… we can do much better

Reply:
should be addressed in 96838e285ea69326c86147ac443a59e5d4773118